bokehjs/src/coffee/embed.ts

```diff
@@ -23,31 +23,42 @@ export interface RenderItem {
   notebook_comms_target?: any
 }
 
-declare interface CommMessage {
+export declare interface CommMessage {
   buffers: DataView[],
   content: {
     data: string
   }
 }
 
-declare interface Comm {
+export declare interface Comm {
   target_name: string
   on_msg: (msg: CommMessage) => void
+  onMsg: (this: Document, receiver: Receiver, comm_msg: CommMessage) => void
+}
+
+export declare interface Kernel {
+  comm_manager: {
+    comms: {[key: string]: Promise<Comm>},
+    register_target: (target: string, fn: (comm: Comm) => void) => void,
+  },
+  _comms: {[key: string]: Promise<Comm>},
+  registerCommTarget: (target: string, fn: (comm: Comm) => void) => void,
 }
 
 declare interface Jupyter {
   notebook: {
-    kernel: undefined | {
-      comm_manager: {
-        comms: {[key: string]: Promise<Comm>},
-        register_target: (target: string, fn: (comm: Comm) => void) => void,
-      }
-    }
+    kernel: Kernel | undefined
   }
 }
 
+
 declare var Jupyter: Jupyter | undefined
 
+export const kernels: {[key: string]: Kernel} = {}
+
+// Exported to allow external libraries to define custom message handlers
+export {Receiver}
```

Comment:
Is this necessary? It's already exported from `protocol` ?

Reply:
Afaik, ``protocol`` is not exported however. It's really not needed for Bokeh itself but it would be really great if we could make it available to other libraries (e.g. HoloViews) somehow. Don't mind at all where, as long as I can access it via the ``Bokeh`` object somehow. I'm planning to write a small example using flask and socketIO to demonstrate how to use it independently of the notebook and bokeh server contexts.

Reply:
FWIW I'd rather export `protocols` so that people can grab it from there, than rely on a transitive import like this

Reply:
Sounds good, I do think these are useful.